src/c++/vdo/base/vdo.c

```diff
@@ -500,7 +940,48 @@ static int initialize_vdo(struct vdo *vdo, struct device_config *config,
 	vdo_initialize_completion(&vdo->admin.completion, vdo, VDO_ADMIN_COMPLETION);
 	init_completion(&vdo->admin.callback_sync);
 	mutex_init(&vdo->stats_mutex);
-	result = read_geometry_block(vdo);
+
+	// Check to see whether we should format the device
+	result = vdo_is_formatted(vdo, &formatted);
+	if (result != VDO_SUCCESS) {
+		*reason = "Could not determine whether layout is empty";
+		return result;
+	}
+
+	// This is here to distiguish between regular error and not formattable.
+	if (formatted == 0) {
+		/* Get size of future indexer */
+		block_count_t index_blocks;
+
+		result = compute_index_blocks(config, &index_blocks);
```

Comment:
Maybe a nitpick, but why is there some logic just in this block, and then more logic in the vdo_format() function?

Call a single function that does all the formatting validation and work, so at least this function is easier to read.